homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,143 @@
+def setup(hass, config):
+    """Setup the tts service."""
+    success = False
+
+    descriptions = load_yaml_config_file(
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    tts_dir = hass.config.path(DOMAIN)
+
+    if not os.path.exists(tts_dir):
+        os.makedirs(tts_dir)
+
+    for platform, p_config in config_per_platform(config, DOMAIN, _LOGGER):
+        tts_implementation = bootstrap.prepare_setup_platform(
+            hass, config, DOMAIN, platform)
+
+        if tts_implementation is None:
+            _LOGGER.error("Unknown tts service specified.")
+            continue
+
+        tts_engine = tts_implementation.get_engine(hass, p_config)
+
+        if tts_engine is None:
+            _LOGGER.error("Failed to initialize tts service %s",
+                          platform)
+            continue
+
+        def generate_speech(tts_service, platform_name, call):
+            """Handle sending tts message service calls."""
+            text = call.data.get(ATTR_TEXT)
+            entity_id = call.data.get(ATTR_ENTITY_ID)
+            allow_cached_file = call.data.get(ATTR_ALLOW_CACHE, True)
+            play_file = call.data.get(ATTR_PLAY, True)
+            language = call.data.get(ATTR_LANGUAGE, config.get(ATTR_LANGUAGE,
+                                                               "en-us"))
+            rate = call.data.get(ATTR_RATE, config.get(ATTR_RATE, 0))
+            codec = call.data.get(ATTR_CODEC, config.get(ATTR_CODEC, "MP3"))
+            audio_format = call.data.get(ATTR_FORMAT,
+                                         config.get(ATTR_FORMAT,
+                                                    "44khz_16bit_stereo"))
+
+            hashed_text = hashlib.sha1(text.encode('utf-8')).hexdigest()
+
+            full_filename = "{}-{}-{}.{}".format(hashed_text, language,
+                                                 platform_name,
+                                                 codec.lower())
+
+            file_path = os.path.join(tts_dir, full_filename)
+
+            if text is None:
+                _LOGGER.error(
+                    'Received call to %s without attribute %s',
+                    call.service, ATTR_TEXT)
+                return
+
+            if (os.path.isfile(file_path)) is False or \
+               (allow_cached_file is False):
+                try:
+                    tts_service.speak(file_path, text, language, rate, codec,
+                                      audio_format)
+                # pylint: disable=bare-except
+                except:
+                    _LOGGER.error('Error trying to speak using %s',
+                                  platform_name)
+                    return
+
+            if play_file is True or entity_id is None:
+                content_url = "{}/api/tts/{}".format(hass.config.api.base_url,
+                                                     full_filename)
+                hass.services.call(core.DOMAIN, SERVICE_PLAY_MEDIA, {
```

Comment:
This service doesn't exist. Also, for calling services, please use the service helpers that components provide
```python
import media_player from homeassistant.components
media_player.play_media(hass, media_type, content_url, entity_id)
```

You want to probably have media_type be passed in. Each component treats it differently, like for Cast it should be buffered.